src/components/state-lookup.ts

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2020 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import Bluebird from "bluebird";
+import PQueue from "p-queue";
+
+interface StateLookupOpts {
+    client: any;
+    stateLookupConcurrency: number;
+    eventTypes?: string[];
+}
+
+interface StateLookupRoom {
+    syncPromise: Promise<StateLookupRoom>;
+    syncComplete: boolean;
+    events: {
+        [eventType: string]: {
+            [stateKey: string]: StateLookupEvent;
+        }
+    }
+}
+
+interface StateLookupEvent {
+    room_id: string;
+    state_key: string;
+    type: string;
+    event_id: string;
+}
+
+const RETRY_STATE_IN_MS = 3000;
+const DEFAULT_STATE_CONCURRENCY = 4;
+
+export default class StateLookup {
+    private _client: any;
+    private eventTypes: {[eventType: string]: boolean} = {};
+    private dict: { [roomId: string]: StateLookupRoom } = {};
+    private lookupQueue: PQueue;
+
+    /**
+     * Construct a new state lookup entity.
+     *
+     * This component stores state events for specific event types which can be
+     * queried at a later date. This component will perform network requests to
+     * fetch the current state for a given room ID. It relies on
+     * {@link StateLookup#onEvent} being called with later events in order to
+     * stay up-to-date. This should be connected to the <code>onEvent</code>
+     * handler on the {@link Bridge}.
+     * @constructor
+     * @param {Object} opts Options for this constructor
+     * @param {MatrixClient} opts.client Required. The client which will perform
+     * /state requests.
+     * @param {string[]} opts.eventTypes The state event types to track.
+     * @throws if there is no client.
+     */
+    constructor (opts: StateLookupOpts) {
+        if (!opts.client) {
+            throw new Error("client property must be supplied");
+        }
+
+        this.lookupQueue = new PQueue({
+            concurrency: opts.stateLookupConcurrency || DEFAULT_STATE_CONCURRENCY,
+        });
+
+        this._client = opts.client;
+        (opts.eventTypes || []).forEach((t) => {
+            this.eventTypes[t] = true;
+        });
+    }
+
+    /**
+     * Get a stored state event.
+     * @param {string} roomId
+     * @param {string} eventType
+     * @param {string=} stateKey If specified, this function will return either
```

Comment:
What does `string=` mean and why are these type annotation kept in addition with TypeScript?

Reply:
For now it's just so the docs continue to be generated successfully. In a later PR we can look to clean these up.